packages/terra-disclosure-manager/src/DisclosureManager.jsx

```diff
@@ -0,0 +1,351 @@
+
+    return newState;
+  }
+
+  constructor(props) {
+    super(props);
+
+    this.resolveDismissPromise = this.resolveDismissPromise.bind(this);
+    this.resolveDismissChecksInSequence = this.resolveDismissChecksInSequence.bind(this);
+
+    this.safelyCloseDisclosure = this.safelyCloseDisclosure.bind(this);
+    this.generatePopFunction = this.generatePopFunction.bind(this);
+
+    this.renderContentComponents = this.renderContentComponents.bind(this);
+    this.renderDisclosureComponents = this.renderDisclosureComponents.bind(this);
+
+    this.openDisclosure = this.openDisclosure.bind(this);
+    this.pushDisclosure = this.pushDisclosure.bind(this);
+    this.popDisclosure = this.popDisclosure.bind(this);
+    this.closeDisclosure = this.closeDisclosure.bind(this);
+    this.requestDisclosureFocus = this.requestDisclosureFocus.bind(this);
+    this.releaseDisclosureFocus = this.releaseDisclosureFocus.bind(this);
+    this.maximizeDisclosure = this.maximizeDisclosure.bind(this);
+    this.minimizeDisclosure = this.minimizeDisclosure.bind(this);
+
+    // These cached functions are stored outside of state to prevent unnecessary rerenders.
+    this.dismissChecks = {};
+    this.onDismissResolvers = {};
+
+    this.state = {
+      disclosureIsOpen: false,
+      disclosureIsFocused: true,
+      disclosureIsMaximized: false,
+      disclosureSize: 'small',
+      disclosureComponentKeys: [],
+      disclosureComponentData: {},
+    };
+  }
+
+  openDisclosure(data) {
+    this.setState({
+      disclosureIsOpen: true,
+      disclosureIsFocused: true,
+      disclosureSize: data.size || availableDisclosureSizes.SMALL,
+      disclosureComponentKeys: [data.content.key],
+      disclosureComponentData: {
+        [data.content.key]: {
+          key: data.content.key,
+          name: data.content.name,
+          props: data.content.props,
+          component: data.content.component,
+        },
+      },
+    });
+  }
+
+  pushDisclosure(data) {
+    const newState = DisclosureManager.cloneDisclosureState(this.state);
+
+    newState.disclosureComponentKeys.push(data.content.key);
+    newState.disclosureComponentData[data.content.key] = {
+      key: data.content.key,
+      name: data.content.name,
+      props: data.content.props,
+      component: data.content.component,
+    };
+
+    this.setState(newState);
+  }
+
+  popDisclosure() {
+    const newState = DisclosureManager.cloneDisclosureState(this.state);
+
+    newState.disclosureComponentData[newState.disclosureComponentKeys.pop()] = undefined;
+
+    this.setState(newState);
+  }
+
+  closeDisclosure() {
+    this.setState({
+      disclosureIsOpen: false,
+      disclosureIsFocused: false,
+      disclosureIsMaximized: false,
+      disclosureSize: availableDisclosureSizes.SMALL,
+      disclosureComponentKeys: [],
+      disclosureComponentData: {},
+    });
+  }
+
+  requestDisclosureFocus() {
+    this.setState({
+      disclosureIsFocused: true,
```

Comment:
Anyway to avoid incurring a re-render with this update? Unless I'm mistaken, the lifecycle would be at the very least 
pushDisclosure and popDisclosure? 

Could how provide info on how many renders occur for the average disclosure?

Reply:
There is only one state change for a given disclosure manager event, so there should only be one render at the DisclosureManager level. However, both the children and disclosure content are rendered, regardless of where the disclosure is actually happening.

I think we could be more performant by caching the built child and disclosure components in state; however, doing so would make a complicated implementation even more complicated. Let's optimize in the future if we identify bottlenecks. The important thing is that we have (passive) ways of doing so.